### PR TITLE
add flux-dump and flux-restore

### PR DESCRIPTION
**Pull request by garlick**
As proposed in #4202, here are dump/restore utilities that walk KVS metadata using the content interface so they can be used when the KVS is not loaded.  They use [libarchive](https://www.libarchive.org/) to support various portable archive formats.

Still playing with these and plan to add tests and man pages but wanted to get this checkpointed.

### CODE REVIEW

**Comment by garlick** [comment]
Some numbers from my test cluster, FWIW:
- 22094 inactive jobs
- 1234 MB content.sqlite
- 122s  to load job manager
- 102s to load job-list

With flux stopped I ran this script, which took 5m+57s to run
```bash
#!/bin/bash
SQLITE_FILE=/net/frankenfurter/pool/scratch/flux/content.sqlite

reader() {
        ./flux start -o,-Sbroker.rc1_path= \
                -o,-Sbroker.rc3_path=\
                -o,-Scontent.backing-path=$SQLITE_FILE.bak \
                bash -c "\
                        flux module load content-sqlite && \
                        flux dump --checkpoint - \
                "
}

writer() {
        ./flux start -o,-Sbroker.rc1_path= \
                -o,-Sbroker.rc3_path= \
                -o,-Scontent.backing-path=$SQLITE_FILE \
                bash -c "\
                        flux module load content-sqlite && \
                        flux restore -v --checkpoint - \
                "
}

mv $SQLITE_FILE $SQLITE_FILE.bak
reader | writer
```
Afterwards:
- 22094 inactive jobs (no change)
- 731 MB content.sqlite (0.59 x)
- 40s  to load job manager (0.32 x)
- 54s to load job-list (0.52 x)

**Comment by grondo** [comment]
> Other things I thought of but felt could be deferred are:

Another idea we had

 * A filter or transform step before restore which would allow - among other things like pruning jobs - restoring to an instance running on a different resource set. This would perhaps allow for easy checkpoint/restore of Flux instances without requiring the instance restart on the same allocation. (We could call it the "everlasting jobstopper")

**Review by chu11 — COMMENTED** [review]
just did a skim, overall looks good, didn't notice much out of place.  just a lot of nits / minor things.

File: doc/man1/flux-restore.rst
Comment (garlick) [reply doc/man1/flux-restore.rst:35]: The level is not used so I'll make it boolean.

File: doc/man1/flux-dump.rst
Comment (garlick) [reply doc/man1/flux-dump.rst:60]: Same here

File: src/common/libkvs/kvs_checkpoint.c
Comment (garlick) [reply src/common/libkvs/kvs_checkpoint.c:116]: Sorry I think you are right.

The thought I was having was that we don't want to artificially prevent a system administrator from reverting a flux-core upgrade that happens to raise  the checkpoint version number here, when the timestamp is probably not changing; however, I think that could be more of an issue for deciding when to bump the version number.  It probably shouldn't be bumped at all unless the new version introduces a breaking change.

So I'll fix that, thanks.

File: doc/man1/flux-dump.rst
Comment (garlick) [reply doc/man1/flux-dump.rst:65]: Dropping the optional arg so the man page will be correct.

File: doc/man1/flux-restore.rst
Comment (garlick) [reply doc/man1/flux-restore.rst:38]: Dropping the optional arg so the man page will be correct.

**Comment by garlick** [comment]
Just pushed an update with changes squashed (hope that's OK):
- made `--verbose` and `--snapshot` boolean options consistent with man page
- changed kvs snapshot helpers to use the `kvs-primary` if the key is NULL since that's all we support now
- restored snapshot object version check
- fixed other typos and whitespace issues called out by @chu11
- improved testing to ensure that "valref" objects are covered, and to count blobs before and after a restore to show GC benefit
- print number of keys and blobs restored on `flux-restore` completion
- fixed a  json memory leak in `flux-restore` found with valgrind
- rebased on current master

**Comment by garlick** [comment]
Oops found some formatting errors in the man pages so forced a push to correct those, and while I was in there, make an editing pass for clarity.

**Comment by garlick** [comment]
I'm hitting a surrogate error in @chu11's job-archive sqlite query tool that I reused here to count blobs in the sqlite database, only  in the inception builder.  It looks like it may be fixed by @grondo's pr #4212.  Fortuitous!  I'll try rebasing this PR on that one and see how it goes.

**Comment by grondo** [comment]
Just trying out this branch on my size=90 full system instance on fluke with 40K+ jobs.

One thing that might be nice to have is a progress meter in `flux dump/restore`. I'm not sure if that is even possible, but a command running for many minutes with no visible output is somewhat disconcerting, and `-v, --verbose` is way too much information. Anything in-between might be helpful, even just a count of data dumped or restored?

**Comment by garlick** [comment]
Argh, sorry, just re-pushed again - I had let a fix for #4210 leak onto this branch that was not quite right.
Also rebased on @grondo's latest copy of his branch.

**Comment by garlick** [comment]
Looks like #4212 fixed my problem.  Thanks!

I added a periodic count of dumped keys to stderr to both `flux-dump` and `flux-restore`.  It is disabled if a `-q,--quiet` option is specified, or if `-v,--verbose`. 

Once #4212 gets merged, I'll rebase this on master and push the progress stuff too.

I think the coverage here could be better, but only if we simulate bad  KVS metadata in the store.  That's not going to be fun.  My vote is to save that for another day.

**Comment by garlick** [comment]
OK, re-pushed after rebasing, and with new progress display (overwrites line on stderr every 100 keys dumped/restored) and  `--quiet` option.

**Comment by grondo** [comment]
Nice! The progress display is very nice, IMO.

**Comment by chu11** [comment]
> Since flux-dump generates the dump by interacting directly with the
content store, a dump may be created when the KVS is not loaded, as
long as the --checkpoint option is supplied so it has a starting blobref.
Otherwise, the starting blobref is obtained by requesting the current
root from the KVS.

Saw this in a commit message and made me wonder, would a `--checkpoint-blobref` option to reasonable for `flux -dump`?  Like if a user just saved off the value after calling `flux kvs getroot`?

Was just a thought.

**Comment by chu11** [comment]
Another random thought / idea.  Should the key a user restores to be a symlink?  Or possibly in a different namespace?  I'm not entirely sure if there's a realistic use case.  I'm imagining like a user wants to re-start their instance from a specific point in time?  Just a random brainstorm, feel free to ignore.

File: t/t2807-dump-cmd.t
Comment (chu11) [inline t/t2807-dump-cmd.t:69]: it took me a while to grok this comment and get to how you go to 8.  Perhaps more clear (I think the +x above is a cut & paste error?)

```
rootdir-v0 (initial rootdir) + (rootdir-v1 (add a) + dir a + dir b) + (rootdir-v2 (add y)) + (rootdir-v3 (add x) + valref x) + (rootdir-v4 (add z)) = 8 blobrefs
```

File: t/t2807-dump-cmd.t
Comment (chu11) [inline t/t2807-dump-cmd.t:97]: for consistency to my 0above comment, dir a, dir b, valref x?

File: doc/man1/flux-dump.rst
Comment (chu11) [inline doc/man1/flux-dump.rst:17]: "writes a snapshot"? (plural write)

File: doc/man1/flux-restore.rst
Comment (chu11) [inline doc/man1/flux-restore.rst:66]: idea: could kvs-restore check that the module isn't loaded?

File: doc/man1/flux-restore.rst
Comment (chu11) [inline doc/man1/flux-restore.rst:30]: perhaps want to add (or add an example?) of how dump + restore garbage collects inactive references?  But on other hand it may not be super wise to advertise, as users may abuse it?

**Comment by garlick** [comment]
Thanks for the additional comments @chu11!

On the two non-inlined ideas:

> would a --checkpoint-blobref option to reasonable for flux -dump? Like if a user just saved off the value after calling flux kvs getroot?

I had a `--blobref` option originally so it would be easy to add back in, but it seemed like any gainful use of it would require too much knowledge about how the KVS internals work, e.g. that you'd want a blobref that refers to a directory that you could get from a commit, or other tricky ways.  I thought the more user-friendly option would be `--key` (a directory name), but that was a little harder to implement, or at least I had a hard time figuring it out at the time.   So...maybe a follow on PR if a use case develops?

> Could the user restoring want the key to be a symlink? Or possibly in a different namespace? I'm not entirely sure if there's a realistic use case. I'm imagining like a user wants to re-start their instance from a specific point in time?

It definitely seems like dump/restore from/to a namespace root makes sense. I actually listed that above as something I thought we might want to add later.   I think a symlink can easily be added after the restore, if the idea there is to  add one that could be followed into another namespace from the primary.

Good ideas IMHO, but could be deferred to if/when a usecase develops?

File: t/t2807-dump-cmd.t
Comment (garlick) [reply t/t2807-dump-cmd.t:69]: sorry, those comments weren't updated when the script changed!  I'll update and expand the comment so the casual reader won't be totally mystified!

File: doc/man1/flux-restore.rst
Comment (garlick) [reply doc/man1/flux-restore.rst:66]: Oooh good idea.

File: doc/man1/flux-restore.rst
Comment (garlick) [reply doc/man1/flux-restore.rst:30]: I think it may be best to hold off on that for a bit.  For now, the things that are omitted from the archive are described in `flux-dump(1)`.  If we're successful with using it for garbage collection, we could revisit.

**Comment by garlick** [comment]
Pushed some fixups (squashed) for @chu11's comments.

The test for the kvs being loaded was a great idea - I was able to remove the CAVEATS section from `flux-restore(1)` now that it is automatically detected.

**Comment by garlick** [comment]
Well something is not right here.  When I dump and restore my system at home, I can no longer run jobs.  I get the following error

```
$ flux mini run hostname
0.475s: job.exception type=exec severity=0 failed to create guest ns: No such file or directory
$ sudo flux dmesg
[snip]
2022-03-15T04:32:12.313805Z kvs.err[0]: content_load_completion: flux_content_load_get: No such file or directory
2022-03-15T04:32:12.317248Z job-exec.info[0]: job-exception: id=2317853860560896: failed to create guest ns: No such file or directory
```
So it looks like there is a dangling blobref there somewhere.

**Review by chu11 — APPROVED** [review]
updates look good, LGTM!

**Comment by garlick** [comment]
Thanks @chu11.

Just pushed  some more fixups (squashed) after running down the problem above:
- add work around and test for #4222
- fixed minor typos in job-exec.c, content-sqlite.c
- flush content store after restoring, unless `--no-cache`
- slightly adjust progress output and make `--quiet` really quiet
- cherry-pick @grondo's commit to add libarchive to Dockerfiles and drop temporary `checks/Dockerfile` commit

**Comment by grondo** [comment]
Sorry the fedora33 docker image build or push failed for unknown reason. I had to restart the build so it will be up to an hour before that is fixed.

**Pull request merged**
